src/SourcemapToolkit.SourcemapParser/SourceMap.cs

```diff
@@ -35,18 +36,93 @@ public class SourceMap
 		public List<MappingEntry> ParsedMappings;
 
 		/// <summary>
-		/// Finds the mapping entry for the generated source position. If no exact match is found, it will attempt 
-		/// to return a nearby mapping that should map to the same piece of code.
+		/// Applies the mappings of a sub source map to the current source map
+		/// Each mapping to the supplied source file is rewritten using the supplied source map
+		/// <param name="submap">The submap to apply</param>
+		/// <param name="aSourceFile">The filename of the source file. If not specified, submap's File property will be used</param>
+		/// <returns>A new source map</returns>
 		/// </summary>
-		/// <param name="generatedSourcePosition">The location in generated code for which we want to discover a mapping entry</param>
-		/// <returns>A mapping entry that is a close match for the desired generated code location</returns>
-		public virtual MappingEntry GetMappingEntryForGeneratedSourcePosition(SourcePosition generatedSourcePosition)
+		public SourceMap ApplySourceMap(SourceMap submap, string sourceFile = null)
 		{
-			if (ParsedMappings == null)
+			if (submap == null)
+			{
+				throw new ArgumentNullException(nameof(submap));
+			}
+
+			if (sourceFile == null)
 			{
-				return null;
+				if (submap.File == null)
+				{
+					throw new Exception("ApplySourceMap expects either the explicit source file to the map, or submap's 'file' property");
+				}
+
+				sourceFile = submap.File;
 			}
 
+			SourceMap newSourceMap = new SourceMap
+			{
+				File = this.File,
+				Version = this.Version,
+				Sources = new List<string>(),
+				Names = new List<string>(),
+				ParsedMappings = new List<MappingEntry>()
+            };
+
+			// transform mappings in this source map
+			foreach (MappingEntry mappingEntry in this.ParsedMappings)
+			{
+				MappingEntry newMappingEntry = mappingEntry.Clone();
+
+				if (mappingEntry.OriginalFileName == sourceFile && mappingEntry.OriginalSourcePosition != null)
+				{
+					MappingEntry correspondingEntry = submap.GetMappingEntryForGeneratedSourcePosition(mappingEntry.OriginalSourcePosition);
+
+					if (correspondingEntry != null)
+					{
+						// Copy the mapping
+						newMappingEntry = new MappingEntry
+						{
+							GeneratedSourcePosition = mappingEntry.GeneratedSourcePosition.Clone(),
+							OriginalSourcePosition = correspondingEntry.OriginalSourcePosition.Clone(),
+							OriginalName = correspondingEntry.OriginalName?? mappingEntry.OriginalName,
+							OriginalFileName = correspondingEntry.OriginalFileName?? mappingEntry.OriginalFileName
+						};
+					}
+				}
+
+				// Copy into "Sources" and "Names"
+				string originalFileName = newMappingEntry.OriginalFileName;
+				string originalName = newMappingEntry.OriginalName;
+
+				if (originalFileName != null && !newSourceMap.Sources.Contains(originalFileName))
+				{
+                    newSourceMap.Sources.Add(originalFileName);
```

Comment:
the diff isn't displaying correct indenting, I think it has something to do with the mixed tabs and spaces. the indenting is correct in my text editor